cmd/get_logs.go

```diff
@@ -0,0 +1,358 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+package cmd
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/Azure/aks-engine/pkg/api"
+	"github.com/Azure/aks-engine/pkg/armhelpers"
+	"github.com/Azure/aks-engine/pkg/engine"
+	"github.com/Azure/aks-engine/pkg/helpers"
+	"github.com/Azure/aks-engine/pkg/i18n"
+	"github.com/leonelquinteros/gotext"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"golang.org/x/crypto/ssh"
+	v1 "k8s.io/api/core/v1"
+)
+
+const (
+	getLogsName             = "get-logs"
+	getLogsShortDescription = "Collect logs and current cluster nodes configuration."
+	getLogsLongDescription  = "Collect deployment logs, running daemons/services logs and current nodes configuration."
+)
+
+type getLogsCmd struct {
+	// user input
+	location               string
+	apiModelPath           string
+	apiserverURI           string
+	linuxSSHPrivateKeyPath string
+	linuxScriptPath        string
+	outputDirectory        string
+	controlPlaneOnly       bool
+	// computed
+	cs               *api.ContainerService
+	locale           *gotext.Locale
+	armClient        armhelpers.AKSEngineClient
+	masterNodes      []v1.Node
+	linuxNodes       []v1.Node
+	linuxSSHConfig   *ssh.ClientConfig
+	windowsNodes     []v1.Node
+	windowsSSHConfig *ssh.ClientConfig
+}
+
+func newGetLogsCmd() *cobra.Command {
+	glc := getLogsCmd{}
+	command := &cobra.Command{
+		Use:   getLogsName,
+		Short: getLogsShortDescription,
+		Long:  getLogsLongDescription,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := glc.validateArgs(); err != nil {
+				cmd.Usage()
+				return errors.Wrap(err, "validating get-logs args")
+			}
+			if err := glc.loadAPIModel(); err != nil {
+				return errors.Wrap(err, "loading API model")
+			}
+			return glc.run()
+		},
+	}
+	command.Flags().StringVarP(&glc.location, "location", "l", "", "Azure location where the cluster is deployed (required)")
+	command.Flags().StringVarP(&glc.apiModelPath, "api-model", "m", "", "path to the generated apimodel.json file (required)")
+	command.Flags().StringVar(&glc.apiserverURI, "apiserver", "", "apiserver endpoint (required)")
```

Comment:
Should we change this to `--ssh-endpoint`? Basically, it's not obvious that the Kubernetes apiserver FQDN would also be an ssh listener endpoint. Maybe:

`command.Flags().StringVar(&glc.sshEndpoint, "ssh-endpoint", "", "FQDN, or IP address, of an ssh listener that has access to ssh into all nodes in the cluster")`

Reply:
I agree, should be more generic. What about `--ssh-bastion` or `--ssh-jumpbox` or similar? 

`endpoint` is probably safer as it is a generic term, but `bastion`/`jumpbox`/`proxy` are terms usually [associated](https://www.google.com/search?q=ssh+proxy+command) to this behavior.

I do not mind either one, pick your favorite.

Reply:
How about `--ssh-host`?

Reply:
perfect